### ISSUE TITLE
Routing selects wrong method in controller

### ISSUE THREAD

**Issue by henrikdahl8240** [OData/WebApi#2075]
I have these definitions in my Entity Model:
```
var function__Bike__SellA = this.EntityType<Bike>()
    .Function("Sell").Returns<SellResult_DTO>();
function__Bike__SellA.Parameter<string>("aaa");
function__Bike__SellA.Parameter<int>("ccc");
function__Bike__SellA.Parameter<string>("ddd");
function__Bike__SellA.Parameter<string>("eee");
function__Bike__SellA.Parameter<string>("fff");
function__Bike__SellA.Parameter<bool>("ggg");
function__Bike__SellA.Parameter<bool>("hhh");
function__Bike__SellA.Parameter<bool>("iii");
function__Bike__SellA.Parameter<bool>("jjj");
function__Bike__SellA.Parameter<bool>("kkk");
function__Bike__SellA.Parameter<bool>("lll");
function__Bike__SellA.Parameter<int>("mmm");
function__Bike__SellA.Parameter<int>("nnn");
function__Bike__SellA.Parameter<int>("ooo");

var function__Bike__SellB = this.EntityType<Bike>()
    .Function("Sell").Returns<SellResult_DTO>();
function__Bike__SellB.Parameter<string>("aaa");
function__Bike__SellB.Parameter<int>("ccc");
function__Bike__SellB.Parameter<string>("open");
function__Bike__SellB.Parameter<string>("ddd");
function__Bike__SellB.Parameter<string>("eee");
function__Bike__SellB.Parameter<string>("fff");
function__Bike__SellB.Parameter<bool>("ggg");
function__Bike__SellB.Parameter<bool>("hhh");
function__Bike__SellB.Parameter<bool>("iii");
function__Bike__SellB.Parameter<bool>("jjj");
function__Bike__SellB.Parameter<bool>("kkk");
function__Bike__SellB.Parameter<bool>("lll");
function__Bike__SellB.Parameter<int>("mmm");
function__Bike__SellB.Parameter<int>("nnn");
function__Bike__SellB.Parameter<int>("ooo");

var function__Bike__SellIdC = this.EntityType<Bike>()
    .Function("Sell").Returns<SellResult_DTO>();
function__Bike__SellIdC.Parameter<string>("aaa");
function__Bike__SellIdC.Parameter<int>("ccc");
function__Bike__SellIdC.Parameter<bool>("ppp");
function__Bike__SellIdC.Parameter<string>("qqq");
function__Bike__SellIdC.Parameter<bool>("rrr");
function__Bike__SellIdC.Parameter<string>("sss");
function__Bike__SellIdC.Parameter<bool>("ttt");
function__Bike__SellIdC.Parameter<string>("uuu");
function__Bike__SellIdC.Parameter<string>("ddd");
function__Bike__SellIdC.Parameter<string>("eee");
function__Bike__SellIdC.Parameter<string>("fff");
function__Bike__SellIdC.Parameter<bool>("ggg");
function__Bike__SellIdC.Parameter<bool>("hhh");
function__Bike__SellIdC.Parameter<bool>("iii");
function__Bike__SellIdC.Parameter<bool>("jjj");
function__Bike__SellIdC.Parameter<bool>("kkk");
function__Bike__SellIdC.Parameter<bool>("lll");
function__Bike__SellIdC.Parameter<int>("mmm");
function__Bike__SellIdC.Parameter<int>("nnn");
function__Bike__SellIdC.Parameter<int>("ooo");

var function__Bike__SellD = this.EntityType<Bike>()
    .Function("Sell").Returns<SellResult_DTO>();
function__Bike__SellD.Parameter<string>("aaa");
function__Bike__SellD.Parameter<int>("ccc");
function__Bike__SellD.Parameter<bool>("vvv");
function__Bike__SellD.Parameter<string>("ddd");
function__Bike__SellD.Parameter<string>("eee");
function__Bike__SellD.Parameter<string>("fff");
function__Bike__SellD.Parameter<bool>("ggg");
function__Bike__SellD.Parameter<bool>("hhh");
function__Bike__SellD.Parameter<bool>("iii");
function__Bike__SellD.Parameter<bool>("jjj");
function__Bike__SellD.Parameter<bool>("kkk");
function__Bike__SellD.Parameter<bool>("lll");
function__Bike__SellD.Parameter<int>("mmm");
function__Bike__SellD.Parameter<int>("nnn");
function__Bike__SellD.Parameter<int>("ooo");

var action__Bike__Sell = this.EntityType<Bike>()
    .Action("Sell").Returns<SellResult_DTO>();
action__Bike__Sell.CollectionParameter<string>("xxx");
action__Bike__Sell.Parameter<bool>("ggg");
action__Bike__Sell.Parameter<bool>("hhh");
action__Bike__Sell.Parameter<bool>("iii");
action__Bike__Sell.Parameter<bool>("jjj");
action__Bike__Sell.Parameter<bool>("kkk");
action__Bike__Sell.Parameter<bool>("lll");
action__Bike__Sell.Parameter<int>("mmm");
action__Bike__Sell.Parameter<int>("nnn");
action__Bike__Sell.Parameter<int>("ooo");

```

I have these definitions in my controller:

```
public partial class BikesController
{
    [HttpGet]
    [EnableQuery]
    public async Task<ActionResult<SellResult_DTO>> Sell([FromODataUri] Guid key, string aaa, int ccc, string ddd, string eee, string fff, bool ggg, bool hhh, bool iii, bool jjj, bool kkk, bool lll, int mmm, int nnn, int ooo, ODataActionParameters parameters, CancellationToken cancellationToken)
    {
        // ...
    }

    [HttpGet]
    [EnableQuery]
    public async Task<ActionResult<SellResult_DTO>> Sell([FromODataUri] Guid key, string aaa, int ccc, string open, string ddd, string eee, string fff, bool ggg, bool hhh, bool iii, bool jjj, bool kkk, bool lll, int mmm, int nnn, int ooo, ODataActionParameters parameters, CancellationToken cancellationToken)
    {
        // ...
    }

    [HttpGet]
    [EnableQuery]
    public async Task<ActionResult<SellResult_DTO>> Sell([FromODataUri] Guid key, string aaa, int ccc, bool ppp, string qqq, bool rrr, string sss, bool ttt, string uuu, string ddd, string eee, string fff, bool ggg, bool hhh, bool iii, bool jjj, bool kkk, bool lll, int mmm, int nnn, int ooo, ODataActionParameters parameters, CancellationToken cancellationToken)
    {
        // ...
    }

    [HttpGet]
    [EnableQuery]
    public async Task<ActionResult<SellResult_DTO>> Sell([FromODataUri] Guid key, string aaa, int ccc, bool vvv, string ddd, string eee, string fff, bool ggg, bool hhh, bool iii, bool jjj, bool kkk, bool lll, int mmm, int nnn, int ooo, ODataActionParameters parameters, CancellationToken cancellationToken)
    {
        // ...
    }

    [HttpPost]
    [ODataRoute("Bikes({key})/" + ODataModelBuilders.Constants.functionAndActionNamespace + ".Sell")]
    public async Task<ActionResult<SellResult_DTO>> Sell([FromODataUri] Guid key, ODataActionParameters parameters, CancellationToken cancellationToken)
    {
        // ...
    }
}

```

I have this query, which in my situation also includes a JSON body holding the parameters for the action:

`POST http://localhost/xxx/odata-internal/Bikes(028e8421-77ef-46d3-7509-08d7bce55a10)/MyServices.Sell()
`
Actions should always be invoked by a POST. It is quite easy to figure out, which method should serve the request, the last one as it is the only method decorated with `[HttpPost]`. Unfortunately it's the very first Sell method, which is actually decorated with `[HttpGet]`, which gets invoked.

Removing the attribute `[ODataRoute("Bikes({key})/MyServices.Sell")]` makes no difference.

My `ConfigureServices(IServiceCollection services)` looks like this:

```
public void ConfigureServices(IServiceCollection services)
{
    services.AddDbContext<Bike_DbContext>(options =>
    {
        options.UseLazyLoadingProxies();
        options.UseSqlServer(Configuration.GetConnectionString("Bike_DbContext"), opts =>
        {
            opts.CommandTimeout((int)TimeSpan.FromMinutes(10).TotalSeconds);
            opts.MigrationsAssembly(typeof(MigrationMarker).Assembly.GetName().Name);
        });
    });

    services.AddControllers(mvcOptions =>
    {
        mvcOptions.EnableEndpointRouting = false;
        mvcOptions.AllowEmptyInputInBodyModelBinding = true;
        mvcOptions.MaxIAsyncEnumerableBufferLimit = 100000;
    });

    services.AddIdentityCore<ApplicationUser>()
        .AddEntityFrameworkStores<Bike_DbContext>();

    services.AddAuthentication("Bearer")
        .AddJwtBearer("Bearer", options =>
        {
            options.Authority = "http://localhost/Bike.OAuthServerCore";
            options.RequireHttpsMetadata = false;

            options.Audience = Component.OAuthServer.Constants.Scope_BikeApi_Name;
        });

    services.AddAuthorization(options =>
    {
        // ...
    });

    services.AddOData();
}

```

My `Configure(IApplicationBuilder app, IWebHostEnvironment env)` looks like this:

```
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{
    if (env.IsDevelopment())
    {
        app.UseDeveloperExceptionPage();
    }

    app.UseCors(MyAllowAnyOrigin);

    app.UseRouting();

    app.UseAuthentication();

    app.UseAuthorization();

    //app.UseEndpoints(endpoints => // (HD) 20200213 Commented out according to https://devblogs.microsoft.com/odata/experimenting-with-odata-in-asp-net-core-3-1/.
    //{
    //    endpoints.MapControllers();
    //});

    app.UseMvc(routeBuilder =>
    {
        routeBuilder.MapODataServiceRoute("odata", "odata-internal", GetEdmModel());
    });
}

```

It all worked perfectly using Microsoft.AspNet.OData 6.1.0 served by .NET Framework.

Why doesn't it work, what should be done to make routing go to the right method?

Why is a method decorated with `[HttpGet]` selected by the router for an action?

Isn't method overloading supported anymore?

Where is it possible to read specifics on these routing matters using Microsoft.AspNetCore.OData 7.3.0?

I am using .NET Core 3.1, Visual Studio 16.4.5 and Windows 10 Pro.

**Comment by xuzhg** [comment]
@henrikdahl8240 Thanks for reporting. This is known issue that ODataActionSelector in Core version  simply select the first matched candidates if there are multiple matched.

The matching process is simple using the method name. We'd like to improve the action selector. 
@habbes Did your PR improve this?

**Comment by henrikdahl8240** [comment]
OK, it clarifies everything, of course.

Is it possible to read about e.g. the differences between 6.1.0 and 7.3.0 on Core somewhere or perhaps not, because it's obviously not everything which is documented? Personally I found

> Where, [FromBody] is necessary in ASP.NET Core.

in your article at https://devblogs.microsoft.com/odata/asp-net-core-odata-now-available/ very valuable.

Why is there actually the attribute `[ODataRoute("CreateMovie")]` put in the action sample for Core at https://github.com/OData/ODataSamples/blob/master/WebApiCore/ODataActionSample/ODataActionSample/Controllers/MoviesController.cs because I thought I might be missing something?

Yesterday I also put some other issues regarding casting, which don't work. May I e-mail you a solution, which documents that so you may take a look at it? I suppose it matches your article https://devblogs.microsoft.com/odata/select-enhancement-in-asp-net-core-odata/, which I also found very valuable.

**Comment by habbes** [comment]
@xuzhg I created a project that reproduces this issue and tested it with my PR. Unfortunately my PR does not solve this, it returns a 404 error instead. The reason my PR fails in this example is because of the extra `CancellationToken` parameter. Because the cancellation token is not part of the arguments passed in the URL, it thinks it's not meant for this action and returns a 404. That sounds like a flaw in my PR though. Without the `CancellationToken` it would have returned the correct method.

**Comment by henrikdahl8240** [comment]
According to https://www.thinktecture.com/en/asp-net/aspnet-core-in-production-graceful-shutdown-and-reacting-to-aborted-requests/ it seems like it should be correct in ASP.NET Core to include a formal parameter of type CancellationToken. You may see the example at the end.

Quoted code:
```
[HttpGet("MyAction")]
public async Task<IActionResult> MyAction(CancellationToken token)
{
    return await DoSomethingAsync(token);
}
```
I would dare to claim, that your PR should ensure, that the same contract between the signature of an action method and the ASP.NET runtime is being maintained, i.e. if the ASP.NET Runtime offers the idea of passing an argument of type CancellationToken, then your PR must also enable that. Otherwise, there may be routed to the right method, but it may be difficult to cancel it, e.g. when the user agent, for instance a browser, closes the connection when refreshing the page, navigaties to another page etc.

In a production situation it's obviously important to be able to clean up in response to decisions made by the user.

**Comment by habbes** [comment]
@henrikdahl8240 I agree with your assessment, it's my PR that needs to be improved to take this into account. I wonder whether there are other AspNet Core routing conventions and contracts that my PR should be aware of that I may have missed.

**Comment by henrikdahl8240** [comment]
Yes, but couldn't it be a reasonable assumption, that this is a standard problem, which the ASP.NET Core Runtime has a standard solution for and you should be able to use that? E.g. to request a service using dependency injection like "IInvokerWithAllBonusArguments" having one method like Invoke("some way of designating the right method")? In this way it could also be future proof, when arguments will come or go. This would compare to the idea of the general constructor, which may inject from the dependency injector when instantiating some new class. Doesn't it sound reasonable?
Unfortunately I have not managed to find a page at https://docs.microsoft.com/ which gives the definitive definition concerning available "bonus arguments", which I suppose we all "need". It's obvious we are interested in knowing the possibilities offered. Have you?

**Comment by habbes** [comment]
@henrikdahl8240 I did some digging to find out how to detect "bonus arguments". Each action parameter is resolved by a model binder, so one could check which `ModelBinder` handles a particular parameter. Each model binder type has an associated `ModelBinderProvider` which are registered by the `AddMvc()` method. You can find the list of built-in model binders here: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.modelbinding.binders?view=aspnetcore-3.1, you can also add custom ones.

From reading the aspnetcore source code, I've found that the easiest way to tell which model binder handles a particular parameter is to use the `IModelBinderFactory.CreateBinder()` (`IModeBinderFactory` can be obtained from DI)

I've made use of this in my new PR #2168, as well as checking for http method attributes declared on the controller actions.

**Comment by henrikdahl8240** [comment]
@habbes I hope it is a sound approach, but I do not have knowledge on the internalt to the extent I may judge it. From computer science point of view it obviously sounds intuitively reasonable.
@xuzhg What do you say, might it be a valid approach?
@xuzhg Do you think, that one of your nightly builds could incorporate this, so it can be used with .Net 5.0 Preview 4, because then I would like to try it, particularly if I do not get the runtime exceptions when it is put under stress, which I have received from beginning using endpoint routing with 7.4.0-beta and from .Net 3.1 until .Net 5.0 Preview 3?
For your information, the exceptions have prevented me from running my migration procedure, so I have not been able to establish a propagated context (with millions of entities managed by EF Core).

**Comment by henrikdahl8240** [comment]
I would just like to inform, that now I have tried using .Net 5.0 Preview 4 and it's the first version where I do not get exceptions from the OData server when putting it under stress after moving to OData 7.4.0 using endpoint routing. I just did not want the comment above to be hanging as the issue does not apply anymore.

**Comment by henrikdahl8240** [comment]
@habbes  Does your PR make all the possibilities at https://docs.microsoft.com/en-us/aspnet/web-api/overview/odata-support-in-aspnet-web-api/odata-routing-conventions available?

Does it for instance make this example available:

POST /entityset(key)/cast/action | /Products(1)/Models.Book/CheckOut | ActionNameOnEntityType or ActionName | CheckOutOnBook
-- | -- | -- | --


For instance you may have CheckOutOnBook and CheckOutOnMagazine if Magazine is also a derived entity of Product.

**Comment by habbes** [comment]
@henrikdahl8240  based on the test suites, I believe that is also supported. But I can try a small repro to confirm that it is the case.

**Comment by henrikdahl8240** [comment]
@habbes It is very kind of you. I hope you will share the result of your findings. It can also be, that the general idea was already supported by the general OData stack for .NET Core. It would also be easier, if there would be corresponding documentation for the .NET Core version, which I have not managed to find.

**Comment by habbes** [comment]
Hi @henrikdahl8240 improvements to action selector were released in 7.5.x, the most recent version of 7.x should address the issue. Feel free to open an issue if you still encounter problems with the new version.
